mlir/tools/rocmlir-gen/rocmlir-gen.cpp

```diff
@@ -2281,6 +2289,103 @@ static func::FuncOp createCpuGemmKernelWithMlir(ModuleOp module,
   return func;
 }
 
+static func::FuncOp createCpuAttentionKernelWithMlir(ModuleOp module,
+                                                     const GenParams &params) {
+  MLIRContext *ctx = module.getContext();
+  OpBuilder builder(ctx);
+  Location loc = module->getLoc();
+
+  // auto elemTypes = typeFromString(inputDataType.getValue(), ctx);
+  SmallVector<Type, 5> argTypes;
+  getAttentionTypes(argTypes, params.types);
+
+  constexpr llvm::StringLiteral cpuKernName("host_naive_attention");
+  auto func = builder.create<func::FuncOp>(
+      loc, cpuKernName, builder.getFunctionType(argTypes, {}));
+
+  Block *block = func.addEntryBlock();
+  builder.setInsertionPointToStart(block);
+
+  auto extendMemref = [&builder, &loc](mlir::Value value) {
```

Comment:
and we can avoid the lambda here too.

Reply:
I guess it will result in more or less the same code. We need to cast2d tensors to 3d because of the requirements imposed by `tosa::MulOp`. So, a small lambda function will be still needed.

Here is my draft,

```cpp
  auto getAsTensor = [&builder, &loc](mlir::Value value, bool isWritable = false) {
    constexpr bool isRestrict{true}; 
    auto origTensor = builder.create<bufferization::ToTensorOp>(loc, value, isRestrict, isWritable);
    auto origShape = origTensor.getType().getShape();

    std::vector<int64_t> expShape(origShape.size() + 1);
    std::copy(origShape.begin(), origShape.end(), expShape.begin() + 1);
    expShape[0] = 1;

    return builder.create<tosa::ReshapeOp>(loc, origTensor, expShape);
  };

  auto queriesTensor = getAsTensor(block->getArgument(0));
  auto keysTensor = getAsTensor(block->getArgument(1));
  auto valuesTensor = getAsTensor(block->getArgument(2));
  auto scaleTensor = getAsTensor(block->getArgument(3));

  auto qkTensor = builder.create<tosa::MatMulOp>(loc, scaleTensor.getType(),
                                                 queriesTensor, keysTensor);

  auto intermediateTensorType = scaleTensor.getType();
  auto sqkTensor = builder.create<tosa::MulOp>(
      loc, intermediateTensorType, qkTensor, scaleTensor, /*shift=*/0);

  // compute type of the reduced tensor
  auto intermediateTensorShape = intermediateTensorType.getShape();

  SmallVector<int64_t, 5> reducedTensorShape(intermediateTensorShape.size());
  std::copy(intermediateTensorShape.begin(), intermediateTensorShape.end(),
            reducedTensorShape.begin());
  constexpr int64_t reductionAxis{2};
  assert(intermediateTensorShape.size() >= reductionAxis);
  reducedTensorShape[reductionAxis] = 1;

  auto reducedTensorType = mlir::RankedTensorType::get(
      reducedTensorShape, intermediateTensorType.getElementType());

  auto sqkMaxs = builder.create<tosa::ReduceMaxOp>(loc, reducedTensorType,
                                                   sqkTensor, reductionAxis);
  auto normilizedSqkTensor = builder.create<tosa::SubOp>(
      loc, intermediateTensorType, sqkTensor, sqkMaxs);
  auto expsTensor = builder.create<tosa::ExpOp>(loc, intermediateTensorType,
                                                normilizedSqkTensor);

  auto expsSums = builder.create<tosa::ReduceSumOp>(loc, reducedTensorType,
                                                    expsTensor, reductionAxis);
  auto invExpsSums =
      builder.create<tosa::ReciprocalOp>(loc, reducedTensorType, expsSums);
  auto softmaxTensor = builder.create<tosa::MulOp>(
      loc, intermediateTensorType, expsTensor, invExpsSums, /*shift=*/0);
  auto resultTensor = builder.create<tosa::MatMulOp>(
      loc, valuesTensor.getType(), softmaxTensor, valuesTensor);

  auto outputTensor = getAsTensor(block->getArgument(4), true);
  auto type = outputTensor.getType();
  auto memrefType = MemRefType::get(type.getShape(), type.getElementType());

  auto outputMemref = builder.create<bufferization::ToMemrefOp>(loc, memrefType, outputTensor);
  auto resultMemref = builder.create<bufferization::ToMemrefOp>(loc, memrefType, resultTensor);

  builder.create<memref::CopyOp>(loc, resultMemref, outputMemref);

  builder.create<func::ReturnOp>(loc);
  module.push_back(func);
  return func;
```

Please, have a look at the fail of the snippet. The most elegant solution would be

```cpp
  builder.create<bufferization::CopyTensorOp>(loc, resultTensor, outputTensor);
```

But unfortunately it does not work because:

```bash
error: failed to legalize operation 'memref.tensor_store'
    memref.tensor_store %0, %expand_shape_9 : memref<1x1024x32xf32>
    ^
<stdin>:121:5: note: see current operation: "memref.tensor_store"(%13, %14) : (tensor<1x1024x32xf32>, memref<1x1024x32xf32>) -> ()
```

Reply:
It is seems that we are not only ones facing the same problem: https://discourse.llvm.org/t/memref-tensor-store-vs-bufferization-to-memref-memref-copy/60887

Reply:
I checked `memref::TensorStoreOp` and it also didn't work

Reply:
I thought it was a bit simpler without having to deal with reassoc idxs (which is essentially what tosa.reshape lowering does).
If you feel that is indifferent, I wont object.